### PR TITLE
Propagate trace from incoming HTTP and to outgoing HTTP

### PR DESCRIPTION
**Pull request by Patrick0308**
## Feature or Problem

Propagate trace from incoming HTTP and to outgoing HTTP

## Related Issues

#3705

### CODE REVIEW

**Comment by Patrick0308** [comment]
```
warning: unused import: `http::HeaderMap`
  --> crates/provider-http-server/src/lib.rs:37:5
   |
37 | use http::HeaderMap;
   |     ^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

error[E0308]: mismatched types
   --> crates/provider-http-server/src/lib.rs:154:30
    |
154 |             &HeaderExtractor(req.headers()),
    |              --------------- ^^^^^^^^^^^^^ expected `http::header::map::HeaderMap`, found `http::HeaderMap`
    |              |
    |              arguments to this struct are incorrect
    |
    = note: `http::HeaderMap` and `http::header::map::HeaderMap` have similar names, but are actually distinct types
```

I use the old version of opentelemetry-http, but  opentelemetry-http dependent the old http version. Maybe we should update all opentelemetry sdk to latest, the opentelemetry-http dependent the newer opentelemetry sdk which is incompatible.

**Comment by Patrick0308** [comment]
I choose not to use the opentelemetry-http public library.

**Review by brooksmtownsend — APPROVED** [review]
@Patrick0308 great work here, this looks nice and simple!

Do you have an example trace that you tested this with that you could share? I can give it a shot locally if not, just want to confirm that this worked as expected 😄 In the meantime, going to request that another @wasmCloud/observability-maintainers gives this a look before merge

**Comment by brooksmtownsend** [comment]
I don't personally have an issue with the approach taken here and not depending on the upstream library. Perhaps we could use that in the future, but I understand the many different versions of the otel libraries creates difficulties.

File: crates/tracing/src/lib.rs
Comment (joonas) [inline crates/tracing/src/lib.rs:18]: Can we put this behind `#[cfg(feature = "otel")]` as well since it uses OTEL libraries?

**Review by thomastaylor312 — APPROVED** [review]

**Pull request merged**
